lib/queue.js

```diff
@@ -46,255 +41,242 @@ FetchQueue.prototype.add = function(protocol, domain, port, path, depth, callbac
 
     var url = protocol + "://" + domain + (port !== 80 ? ":" + port : "") + path;
 
-    self.exists(protocol, domain, port, path,
-        function(err, exists) {
-            if (err) {
-                return callback(err);
-            }
-
-            if (!exists) {
-                var queueItem = {
-                    url: url,
-                    protocol: protocol,
-                    host: domain,
-                    port: port,
-                    path: path,
-                    depth: depth,
-                    fetched: false,
-                    status: "queued",
-                    stateData: {}
-                };
-
-                self.push(queueItem);
-                callback(null, queueItem);
-            } else {
-                var error = new Error("Resource already exists in queue!");
-                error.code = "DUP";
-
-                callback(error);
-            }
-        });
+    queue.exists(protocol, domain, port, path, function(err, exists) {
+        if (err) {
+            callback(err);
+        } else if (!exists) {
+            var queueItem = {
+                url: url,
+                protocol: protocol,
+                host: domain,
+                port: port,
+                path: path,
+                depth: depth,
+                fetched: false,
+                status: "queued",
+                stateData: {}
+            };
+
+            queue.push(queueItem);
+            callback(null, queueItem);
+        } else {
+            var error = new Error("Resource already exists in queue!");
+            error.code = "DUP";
+
+            callback(error);
+        }
+    });
 };
 
 // Check if an item already exists in the queue...
 FetchQueue.prototype.exists = function(protocol, domain, port, path, callback) {
-    callback = callback && callback instanceof Function ? callback : function() {};
-
+    callback = callback instanceof Function ? callback : noop;
     port = port !== 80 ? ":" + port : "";
 
     var url = (protocol + "://" + domain + port + path).toLowerCase();
 
     if (this.scanIndex[url]) {
         callback(null, 1);
-        return 1;
+    } else {
+        this.scanIndex[url] = true;
+        callback(null, 0);
     }
+};
 
-    this.scanIndex[url] = true;
-    callback(null, 0);
-    return 0;
-
+// Get item from queue
+FetchQueue.prototype.get = function(index, callback) {
+    callback = callback instanceof Function ? callback : noop;
```

Comment:
Hmm you might be right that it's more convenient for the developer to get back an error when the index parameter is out of range. I figured that maybe the developer should keep track of the range themselves, but if we verify the range, then the developer can be sure that the second callback parameter is always a queue item so long as there's no error. I'll revert this!